api/src/main/java/edu/wpi/first/shuffleboard/api/sources/recording/Recording.java

```diff
@@ -30,77 +38,132 @@ public void append(TimestampedData data) {
+  }
+
+  /**
+   * Takes a snapshot of this recording. The snapshot is immutable and thread-safe.
+   *
+   * @return a snapshot of this recording
+   */
+  public Snapshot takeSnapshot() {
+    return ThreadUtils.withLock(lock.readLock(), () -> Snapshot.of(this));
+  }
+
+  /**
+   * Takes a snapshot of this recording, then clears the data in this recording. This should <strong>ONLY</strong>
+   * be called by the recording mechanism.
+   *
+   * @return a snapshot of this recording
+   */
+  @SuppressWarnings("PMD.DefaultPackage") // This should only be used by the Serialization class in the same package
+  Snapshot takeSnapshotAndClear() {
+    return ThreadUtils.withLock(lock.writeLock(), () -> {
+      var snapshot = Snapshot.of(this);
+      data.clear();
+      markers.clear();
+      return snapshot;
+    });
   }
 
   @SuppressWarnings("JavadocMethod")
   public List<String> getSourceIds() {
-    synchronized (modificationLock) {
-      return sourceIds;
-    }
+    return ThreadUtils.withLock(lock.readLock(), () -> ImmutableList.copyOf(sourceIds));
   }
 
   @SuppressWarnings("JavadocMethod")
   public TimestampedData getFirst() {
-    synchronized (modificationLock) {
-      return first;
-    }
+    return ThreadUtils.withLock(lock.readLock(), () -> first);
   }
 
   @SuppressWarnings("JavadocMethod")
   public TimestampedData getLast() {
-    synchronized (modificationLock) {
-      return last;
-    }
+    return ThreadUtils.withLock(lock.readLock(), () -> last);
   }
 
   /**
    * Gets the length of this recording in milliseconds. Recordings wth 0 or 1 data points have a length of 0.
    */
   public long getLength() {
-    synchronized (modificationLock) {
+    return ThreadUtils.withLock(lock.readLock(), () -> {
       if (first == null || last == null) {
-        return 0;
+        return 0L;
       } else {
         return last.getTimestamp() - first.getTimestamp();
       }
-    }
+    });
   }
 
   @Override
-  public boolean equals(Object obj) {
-    synchronized (modificationLock) {
-      if (this == obj) {
-        return true;
-      }
-      if (obj == null || getClass() != obj.getClass()) {
-        return false;
-      }
+  public String toString() {
+    return ThreadUtils.withLock(lock.readLock(), () -> "Recording(data=" + data + ", markers=" + markers + ")");
+  }
 
-      Recording that = (Recording) obj;
+  /**
+   * A snapshot of the state of a recording. Snapshots are immutable and thread-safe.
+   */
+  public static final class Snapshot {
+    private final ImmutableList<TimestampedData> data;
+    private final ImmutableList<Marker> markers;
 
-      return this.data.equals(that.data);
+    @SuppressWarnings("PMD.DefaultPackage") // This should only be used by the Serialization class in the same package
+    static Snapshot of(Recording recording) {
+      return new Snapshot(
+          ImmutableList.copyOf(recording.data),
+          ImmutableList.copyOf(recording.markers)
+      );
```

Comment:
I'd recommend indicating that this method itself isn't thread-safe. I know you're calling it in a thread safe way, but this method alone isn't.
There's also no harm in making it thread safe by calling this constructor in a `ThreadUtils.withLock(recording.lock.readLock(), () -> ....` block.
You can graph the read lock from within the write lock just FYI.

Reply:
I don't think this needs to be explicitly threadsafe. It'll also trigger PMD or spotbugs about referencing a private field in a parent class

Reply:
> It'll also trigger PMD or spotbugs about referencing a private field in a parent class
You're already doing that man...